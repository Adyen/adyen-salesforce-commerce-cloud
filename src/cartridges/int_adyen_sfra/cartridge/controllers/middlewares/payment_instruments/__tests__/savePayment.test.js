/* eslint-disable global-require */
let savePayment;
let res;
let req;

beforeEach(() => {
  const { paymentInstruments } = require('../../index');
  savePayment = paymentInstruments.savePayment;
  jest.clearAllMocks();
  res = { json: jest.fn() };
  req = { currentCustomer: { profile: { customerNo: 'mocked_customerNo' } } };
});

afterEach(() => {
  jest.resetModules();
});

describe('Save Payment', () => {
  it('should do nothing if payment processor is not Adyen', () => {
    const PaymentMgr = require('dw/order/PaymentMgr');
    const server = require('server');
    PaymentMgr.getPaymentMethod.mockImplementation(() => ({
      getPaymentProcessor: jest.fn(() => ({
        getID: jest.fn(() => 'notAdyen')
      })),
      isActive: jest.fn(() => false)
    }));

    savePayment.call({ emit: jest.fn() }, req, res, jest.fn());
    expect(server.forms.getForm).toBeCalledTimes(0);
  });

  it('should fail if zeroAuth has error', () => {
    const adyenZeroAuth = require('*/cartridge/scripts/adyenZeroAuth');
    adyenZeroAuth.zeroAuthPayment.mockImplementation(() => ({
      error: true,
    }));
    savePayment.call({ emit: jest.fn() }, req, res, jest.fn());
    expect(res.json.mock.calls).toMatchSnapshot();
  });

  it('should fail if resultCode is not Authorised', () => {
    const adyenZeroAuth = require('*/cartridge/scripts/adyenZeroAuth');
    adyenZeroAuth.zeroAuthPayment.mockImplementation(() => ({
      resultCode: 'Not_Authorised',
    }));
    savePayment.call({ emit: jest.fn() }, req, res, jest.fn());
    expect(res.json.mock.calls).toMatchSnapshot();
  });

  it('should succeed', () => {
    savePayment.call({ emit: jest.fn() }, req, res, jest.fn());
    expect(res.json.mock.calls).toMatchSnapshot();
  });

  it('should return redirectAction and succeed', () => {
    const adyenZeroAuth = require('*/cartridge/scripts/adyenZeroAuth');
    adyenZeroAuth.zeroAuthPayment.mockReturnValue({
      resultCode: 'RedirectShopper',
      action: {
        paymentMethodType: "scheme",
        url: "https://checkoutshopper-test.adyen.com/checkoutshopper/threeDS2.shtml",
        data: {
          MD: "mockMD",
          PaReq: "mockPaReq",
          TermUrl: "https://checkoutshopper-test.adyen.com/checkoutshopMock"},
        method: "POST",
        type: "redirect"
      }
    });
    savePayment.call({ emit: jest.fn() }, req, res, jest.fn());
    expect(res.json.mock.calls).toMatchSnapshot();
  });
});
