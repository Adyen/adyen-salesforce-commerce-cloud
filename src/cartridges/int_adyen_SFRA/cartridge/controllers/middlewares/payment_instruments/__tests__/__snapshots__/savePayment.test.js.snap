// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Save Payment should fail if resultCode is not Authorised 1`] = `
[
  [
    {
      "error": [
        "mocked_error.card.information.error",
      ],
      "success": false,
    },
  ],
]
`;

exports[`Save Payment should fail if zeroAuth has error 1`] = `
[
  [
    {
      "error": [
        "mocked_error.card.information.error",
      ],
      "success": false,
    },
  ],
]
`;

exports[`Save Payment should return redirectAction and succeed 1`] = `
[
  [
    {
      "redirectAction": {
        "data": {
          "MD": "mockMD",
          "PaReq": "mockPaReq",
          "TermUrl": "https://checkoutshopper-test.adyen.com/checkoutshopMock",
        },
        "method": "POST",
        "paymentMethodType": "scheme",
        "type": "redirect",
        "url": "https://checkoutshopper-test.adyen.com/checkoutshopper/threeDS2.shtml",
      },
      "redirectUrl": "["PaymentInstruments-List"]",
      "success": true,
    },
  ],
]
`;

exports[`Save Payment should succeed 1`] = `
[
  [
    {
      "redirectUrl": "["PaymentInstruments-List"]",
      "success": true,
    },
  ],
]
`;
